include/alpaka/queue/cuda_hip/QueueUniformCudaHipRtBase.hpp

```diff
@@ -142,22 +165,133 @@ namespace alpaka
+            {
+                // explicitly copy the shared_ptr so that this method holds the state even when the executing thread
+                // has already finished.
+                const auto pCallbackSynchronizationData
+                    = reinterpret_cast<CallbackSynchronizationData*>(arg)->shared_from_this();
+
+                // Notify the executing thread.
+                {
+                    std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                    pCallbackSynchronizationData->state = CallbackState::notified;
+                }
+                pCallbackSynchronizationData->m_event.notify_one();
+
+                // Wait for the executing thread to finish the task if it has not already finished.
+                std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                if(pCallbackSynchronizationData->state != CallbackState::finished)
+                {
+                    pCallbackSynchronizationData->m_event.wait(
+                        lock,
+                        [pCallbackSynchronizationData]()
+                        { return pCallbackSynchronizationData->state == CallbackState::finished; });
+                }
+            }
+
+            ALPAKA_FN_HOST static auto enqueue(
+                uniform_cuda_hip::detail::QueueUniformCudaHipRtBase<TBlocking>& queue,
+                TTask const& task) -> void
+            {
+                auto pCallbackSynchronizationData = std::make_shared<CallbackSynchronizationData>();
+                ALPAKA_UNIFORM_CUDA_HIP_RT_CHECK(ALPAKA_API_PREFIX(StreamAddCallback)(
+                    queue.getNativeHandle(),
+                    uniformCudaHipRtCallback,
+                    pCallbackSynchronizationData.get(),
+                    0u));
+
+                // We start a new std::thread which stores the task to be executed.
+                // This circumvents the limitation that it is not possible to call CUDA/HIP methods within the CUDA/HIP
+                // callback thread. The CUDA/HIP thread signals the std::thread when it is ready to execute the task.
+                // The CUDA/HIP thread is waiting for the std::thread to signal that it is finished executing the task
+                // before it executes the next task in the queue (CUDA/HIP stream).
+                std::thread t(
+                    [pCallbackSynchronizationData, task]()
+                    {
+                        // If the callback has not yet been called, we wait for it.
+                        {
+                            std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                            if(pCallbackSynchronizationData->state != CallbackState::notified)
+                            {
+                                pCallbackSynchronizationData->m_event.wait(
+                                    lock,
+                                    [pCallbackSynchronizationData]()
```

Comment:
```suggestion
                                    [&pCallbackSynchronizationData]
```

Reply:
I didn't write this - but `pCallbackSynchronizationData` is a pointer, so capturing by value seems reasonable.

Reply:
You are right, it is a raw pointer here.